Orm/Xtensive.Orm/Core/DisposableSet.cs

```diff
@@ -102,11 +103,42 @@ public DisposableSet()
     void IDisposable.Dispose()
     {
       try {
-        if (list==null)
+        if (list==null) {
           return;
+        }
+
         using (var aggregator = new ExceptionAggregator()) {
-          for (int i = list.Count - 1; i >= 0; i--)
+          for (var i = list.Count - 1; i >= 0; i--) {
             aggregator.Execute(d => d.Dispose(), list[i]);
+          }
+
+          aggregator.Complete();
+        }
+      }
+      finally {
+        set = null;
+        list = null;
+      }
+    }
+
+    async ValueTask IAsyncDisposable.DisposeAsync()
+    {
+      try {
+        if (list==null) {
+          return;
+        }
+
+        using (var aggregator = new ExceptionAggregator()) {
+          for (var i = list.Count - 1; i >= 0; i--) {
+            var disposable = list[i];
+            if (disposable is IAsyncDisposable asyncDisposable) {
+              await aggregator.ExecuteAsync(d => d.DisposeAsync(), asyncDisposable).ConfigureAwait(false);
+            }
+            else {
+              aggregator.Execute(d => d.Dispose(), disposable);

```

Comment:
Can throw NRE because of:
https://github.com/DataObjects-NET/dataobjects-net/blob/92849b49d3ef9f712ec17271b7771d3be3c8e4da/Orm/Xtensive.Orm/Core/DisposableSet.cs#L90

Reply:
I checked this and no NRE is not possible here because `Add` method takes care of this case. It just doesn't add `null`s to internal data structures.